exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/FunctionImplementationRegistry.java

```diff
@@ -186,4 +226,105 @@ public boolean isFunctionComplexOutput(String name) {
     return false;
   }
 
+  public RemoteFunctionRegistry getRemoteFunctionRegistry() {
+    return remoteFunctionRegistry;
+  }
+
+  public List<Func> validate(Path path) throws IOException {
+    URL url = path.toUri().toURL();
+    URL[] urls = {url};
+    ClassLoader classLoader = new URLClassLoader(urls);
+    return drillFuncRegistry.validate(path.getName(), scan(classLoader, path, urls));
+  }
+
+  public void register(String jarName, ScanResult classpathScan, ClassLoader classLoader) {
+    drillFuncRegistry.register(jarName, classpathScan, classLoader);
+  }
+
+  public void unregister(String jarName) {
+    drillFuncRegistry.unregister(jarName);
+  }
+
+  /**
+   * Loads all missing functions from remote registry.
+   * Compares list of already registered jars and remote jars, loads missing jars.
+   * Missing jars are stores in local DRILL_UDF_DIR.
+   *
+   * @return true if at least functions from one jar were loaded
+   */
+  public boolean loadRemoteFunctions() {
+    List<String> missingJars = Lists.newArrayList();
+    Registry registry = remoteFunctionRegistry.getRegistry();
+
+    List<String> localJars = drillFuncRegistry.getAllJarNames();
+    for (Jar jar : registry.getJarList()) {
+      if (!localJars.contains(jar.getName())) {
+        missingJars.add(jar.getName());
+      }
+    }
+
+    for (String jarName : missingJars) {
+      try {
+        Path localUdfArea = new Path(new File(getUdfDir()).toURI());
+        Path registryArea = remoteFunctionRegistry.getRegistryArea();
+        FileSystem fs = remoteFunctionRegistry.getFs();
+
+        String sourceName = JarUtil.getSourceName(jarName);
+
+        Path remoteBinary = new Path(registryArea, jarName);
+        Path remoteSource = new Path(registryArea, sourceName);
+
+        Path binary = new Path(localUdfArea, jarName);
+        Path source = new Path(localUdfArea, sourceName);
+
+        fs.copyToLocalFile(remoteBinary, binary);
+        fs.copyToLocalFile(remoteSource, source);
+
+        URL[] urls = {binary.toUri().toURL(), source.toUri().toURL()};
+        ClassLoader classLoader = new URLClassLoader(urls);
+        ScanResult scanResult = scan(classLoader, binary, urls);
+        drillFuncRegistry.register(jarName, scanResult, classLoader);
+      } catch (IOException | FunctionValidationException e) {
+        logger.error("Problem during remote functions load from {}", jarName, e);
+      }
+    }
+
+    return missingJars.size() > 0;
+  }
+
+  private ScanResult scan(ClassLoader classLoader, Path path, URL[] urls) throws IOException {
```

Comment:
First it scans jar for it's marker file which indicates with packages to scan, and if it exists returns scan result which contains list of classes, annotations, packages found in our jar.
Agree, I'll should have added description to method.